todo-app/src/App.js

```diff
@@ -1,10 +1,85 @@
-import React, { Component } from 'react';
+import React, { Component } from 'react'
+import moment from "moment"
+
+import "./style.css"
+import Header from "./components/Header.js"
+import TodoList from "./components/TodoList"
+import myTodoList from "./components/myTodoList.json"
+import AddingContainer from "./components/AddingContainer"
+import DeleteButton from "./components/DeleteButton"
+
 
 class App extends Component {
+  state = {
+    tasks: JSON.parse(localStorage.getItem("todo list")) || myTodoList
+  }
+  componentWillUpdate(nextProps, nextState) {
+    localStorage.setItem("todo list", JSON.stringify(nextState.tasks))
+  }
+  nextId = () => {
+    let uniqueId;
+    const ids = []
+    // eslint-disable-next-line
+    this.state.tasks.map(task => {
+      ids.push(task.id)
+      uniqueId = Math.max(...ids) + 1
+
+    })
+    return uniqueId ? uniqueId : 1
+  }
+  addingTask = (description, deadLine) => {
+    this.setState((previousState) => ({
+      tasks: [
+        {
+          id: this.nextId(),
+          description: description,
+          deadLine: deadLine,
+          done: false,
+          creatingDate: moment().format("llll")
+        },
+        ...previousState.tasks]
+    })
+    )
+  }
+  handleCheckingBox = (taskId) => {
+    const newState = this.state.tasks
+    // eslint-disable-next-line
+    newState.map(task => {
+      // eslint-disable-next-line
+      (taskId !== task.id) ? task : task.done = !task.done
+    })
+    this.setState({ tasks: newState })
```

Comment:
Here, you use `map` as if it was `forEach`. We use `map` to return new array with mapped values. Like if you have a computation that you need to apply it on all the elements on the array. for example: `[1,2,3,4].map( (i) => i*2);` the return of this will be a new array with `[2, 4, 6, 8]`.

In your code, you don't return anything from the `.map` instead you just mutate/change the state before calling *setState*

Try to add this `console.log` before and after `.map`. You'll find that the second `console.log` prints the `this.state` already changed `done` even before calling `setState`!!

```
handleCheckingBox = (taskId) => {
    const newState = this.state.tasks
    console.log('State Before mutation', this.state.tasks)
    // eslint-disable-next-line
    newState.map(task => {
      // eslint-disable-next-line
      (taskId !== task.id) ? task : task.done = !task.done
    });
    console.log('State After mutation', this.state.tasks);
    this.setState({ tasks: newState })
  }
```

As we agreed we should not update or mutate our state directly without *setState*.

The proper solution is as follows

```
handleCheckingBox = (taskId) => {
    const tasks = this.state.tasks
    console.log('State Before mutation', this.state.tasks)
    // eslint-disable-next-line
    const updatedTasks = tasks.map(task => {
      // eslint-disable-next-line
      if (task.id === taskId) {
        return {
          ...task, done: !task.done
        }
      }

      return task;
    });
    console.log('State After mutation', this.state.tasks);
    this.setState({ tasks: updatedTasks })
  }
```

If you check console, you will find both arrays are the same and the `done` property didn't change at all. But it will change after calling `setState`.

**Please don't hesitate to ask me directly if you still don't understand it.**